src/hotspot/share/interpreter/bytecodeTracer.cpp

```diff
@@ -396,8 +388,78 @@ void BytecodePrinter::print_field_or_method(int orig_i, int i, outputStream* st)
+      (tag.is_dynamic_constant() || tag.is_invoke_dynamic())) {
+    print_dynamic(orig_i, i, tag, st);
+  }
 }
 
+void BytecodePrinter::print_dynamic(int orig_i, int i, constantTag tag, outputStream* st) {
+  ConstantPool* constants = method()->constants();
+  int bsm = constants->bootstrap_method_ref_index_at(i);
+  const char* ref_kind = "";
+  switch (constants->method_handle_ref_kind_at(bsm)) {
+  case JVM_REF_getField         : ref_kind = "REF_getField"; break;
+  case JVM_REF_getStatic        : ref_kind = "REF_getStatic"; break;
+  case JVM_REF_putField         : ref_kind = "REF_putField"; break;
+  case JVM_REF_putStatic        : ref_kind = "REF_putStatic"; break;
+  case JVM_REF_invokeVirtual    : ref_kind = "REF_invokeVirtual"; break;
+  case JVM_REF_invokeStatic     : ref_kind = "REF_invokeStatic"; break;
+  case JVM_REF_invokeSpecial    : ref_kind = "REF_invokeSpecial"; break;
+  case JVM_REF_newInvokeSpecial : ref_kind = "REF_newInvokeSpecial"; break;
+  case JVM_REF_invokeInterface  : ref_kind = "REF_invokeInterface"; break;
+  default                       : ShouldNotReachHere();
+  }
+  st->print("  BSM: %s", ref_kind);
+  print_field_or_method(-i, constants->method_handle_index_at(bsm), st);
+  int argc = constants->bootstrap_argument_count_at(i);
+  st->print("  arguments[%d] = {", argc);
+  if (argc > 0) {
+    st->cr();
+    for (int arg_i = 0; arg_i < argc; arg_i++) {
+      int arg = constants->bootstrap_argument_index_at(i, arg_i);
+      st->print("    ");
+      print_constant_nocheck(arg, st);
+    }
+  }
+  st->print_cr("  }");
+  if (tag.is_invoke_dynamic()) {
+    int indy_index = orig_i;
+    int cpc_index = constants->invokedynamic_cp_cache_index(indy_index);
+    print_cpcache_entry(cpc_index, st);
+  } else {
+    // TODO: print info for tag.is_dynamic_constant()
+  }
+}
+
+void BytecodePrinter::print_invoke_handle(int i, outputStream* st) {
+  print_cpcache_entry(ConstantPool::decode_cpcache_index(i), st);
+}
+
+void BytecodePrinter::print_cpcache_entry(int cpc_index, outputStream* st) {
+  ConstantPool* constants = method()->constants();
+  ConstantPoolCacheEntry* cpce = constants->cache()->entry_at(cpc_index);
+  st->print("  ConstantPoolCacheEntry: ");
+  cpce->print(st, cpc_index);
+  if (cpce->bytecode_1() == Bytecodes::_invokehandle ||
+      cpce->bytecode_1() == Bytecodes::_invokedynamic) {
+    constantPoolHandle cph(Thread::current(), constants);
+    Method* m = cpce->method_if_resolved(cph);
+    oop appendix = cpce->appendix_if_resolved(cph);
+    if (m != NULL) {
+      st->print_cr("  Method%s: " INTPTR_FORMAT " %s.%s%s",
+                   m->is_native() ? " (native)" : "",
+                   p2i(m),
+                   m->method_holder()->name()->as_C_string(),
+                   m->name()->as_C_string(), m->signature()->as_C_string());
+    }
+    if (appendix != NULL) {
+      st->print("  appendix: ");
+      appendix->print_on(st);
+    }
+  }
+}
```

Comment:
Line 444-461 should call a function in cpCache like:
void ConstantPoolCacheEntry::print(outputStream* st, int index) const {

If this doesn't have the format you want, you can enhance this function.

Reply:
I moved this code to `ConstantPoolCacheEntry::print` in version [e4afb9f](https://github.com/openjdk/jdk/pull/9957/commits/e4afb9fa235551ef98ccca408aeba4c18bf9904c)